rotkehlchen/exchanges/iconomi.py

```diff
@@ -0,0 +1,287 @@
+
+        headers = {}
+        if authenticated:
+            signature = self._generate_signature(
+                request_type=verb.upper(),
+                request_path=request_path_no_args,
+                timestamp=timestamp,
+            )
+            headers.update({
+                'ICN-SIGN': signature,
+                'ICN-TIMESTAMP': timestamp,
+                'ICN-API-KEY': self.api_key,
+            })
+
+        if data != '':
+            headers.update({
+                'Content-Type': 'application/json',
+                'Content-Length': str(len(data)),
+            })
+
+        log.debug('ICONOMI API Query', verb=verb, request_url=request_url)
+
+        try:
+            response = getattr(self.session, verb)(
+                request_url,
+                data=data,
+                timeout=30,
+                headers=headers,
+            )
+        except requests.exceptions.RequestException as e:
+            raise RemoteError(f'ICONOMI API request failed due to {str(e)}') from e
+
+        try:
+            json_ret = rlk_jsonloads(response.text)
+        except JSONDecodeError as exc:
+            raise RemoteError('ICONOMI returned invalid JSON response') from exc
+
+        if response.status_code not in (200, 201):
+            if isinstance(json_ret, dict) and 'message' in json_ret:
+                raise RemoteError(json_ret['message'])
+
+            raise RemoteError(
+                'ICONOMI api request for {} failed with HTTP status code {}'.format(
+                    response.url,
+                    response.status_code,
+                ),
+            )
+
+        return json_ret
+
+    def validate_api_key(self) -> Tuple[bool, str]:
+        """
+        Validates that the ICONOMI API key is good for usage in Rotki
+        """
+
+        try:
+            self._api_query('get', 'user/balance')
+            return True, ""
+
+        except RemoteError:
+            return False, 'Provided API Key is invalid'
+
+    def query_balances(self, **kwargs: Any) -> Tuple[Optional[Dict[Asset, Dict[str, Any]]], str]:
+        balances = {}
+        try:
+            resp_info = self._api_query('get', 'user/balance')
+        except RemoteError as e:
+            msg = (
+                'ICONOMI API request failed. Could not reach ICONOMI due '
+                'to {}'.format(e)
+            )
+            log.error(msg)
+            return None, msg
+
+        if resp_info['currency'] != 'USD':
+            raise NotImplementedError("API did not return values in USD")
+
+        for balance_info in resp_info['assetList']:
+            ticker = balance_info['ticker']
+            try:
+                asset = iconomi_asset(ticker)
+
+                # There seems to be a bug in the ICONOMI API regarding balance_info['value'].
+                # The value is supposed to be in USD, but is actually returned
+                # in EUR. So let's use the Inquirer for now.
```

Comment:
This is really interesting. Would be nice to see if their support can say anything about this.